fl4health/clients/basic_client.py

```diff
@@ -1206,3 +1276,55 @@ def transform_gradients(self, losses: TrainingLosses) -> None:
             losses (TrainingLosses): The losses object from the train step
         """
         pass
+
+    def save_checkpoint(self) -> None:
+        """
+        Saves checkpoint dict consisting of client name, total steps, lr schedulers,
+            metrics reporter and optimizers state. Method can be overridden to augment saved checkpointed state.
+        """
+
+        assert self.per_round_checkpointer is not None
+
+        ckpt = {
+            "lr_schedulers_state": {key: scheduler.state_dict() for key, scheduler in self.lr_schedulers.items()},
+            "total_steps": self.total_steps,
+            "client_name": self.client_name,
+            "metrics_reporter": self.metrics_reporter,
+            "optimizers_state": {key: opt.state_dict()["state"] for key, opt in self.optimizers.items()},
+        }
+
+        self.per_round_checkpointer.save_checkpoint(ckpt)
+
+        log(INFO, f"Saving client state to checkpoint at {self.per_round_checkpointer.checkpoint_path}")
+
+    def load_checkpoint(self) -> None:
+        """
+        Load checkpoint dict consisting of client name, total steps, lr schedulers, metrics
+            reporter and optimizers state. Method can be overriden to augment loaded checkpointed state.
+        """
+        assert self.per_round_checkpointer is not None and self.per_round_checkpointer.checkpoint_exists()
+
+        ckpt = self.per_round_checkpointer.load_checkpoint()
+
+        assert "lr_schedulers_state" in ckpt and isinstance(ckpt["lr_schedulers_state"], dict)
+        assert "client_name" in ckpt and isinstance(ckpt["client_name"], str)
+        assert "total_steps" in ckpt and isinstance(ckpt["total_steps"], int)
+        assert "metrics_reporter" in ckpt and isinstance(ckpt["metrics_reporter"], MetricsReporter)
+        assert "optimizers_state" in ckpt and isinstance(ckpt["optimizers_state"], dict)
+
+        self.client_name = ckpt["client_name"]
+        self.total_steps = ckpt["total_steps"]
+        self.metrics_reporter = ckpt["metrics_reporter"]
+
+        # Optimizer is updated in setup_client to reference model weights from server
+        # Thus, only optimizer state (per parameter values such as momentum)
+        # should be loaded
+        for opt, state in zip(self.optimizers.values(), ckpt["optimizers_state"].values()):
```

Comment:
This zip makes me slightly nervous, as I'm not 100% confident that dictionary.values() is always going to have the same order between two different dictionaries. I don't think that is guaranteed, but I might be wrong? I would suggest we use `ckpt["optimizers_state"].items()` and then use the key to get the correct optimizer from `self.optimizers`.

Reply:
As of Python 3.7 dictionaries are ordered (ie maintains insertion order and successive calls to values/items/keys will return the same ordering), so I am pretty sure its okay. But I can definitely add a note to change it over so things are more explicit

Reply:
ckpt["optimizers_state"] is generated by iterating through self.optimizers, so it must have the same ordering